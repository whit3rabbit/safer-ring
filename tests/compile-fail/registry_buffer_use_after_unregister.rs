use safer_ring::Registry;
use std::pin::Pin;

fn main() {
    let mut registry = Registry::new();
    let buffer = Pin::new(Box::new([0u8; 1024]));
    let registered_buffer = registry.register_buffer(buffer).unwrap();
    
    // Unregister the buffer
    let _returned_buffer = registry.unregister_buffer(registered_buffer).unwrap();
    
    // Try to use the buffer after unregistration - should not compile
    println!("Buffer size: {}", registered_buffer.size()); //~ ERROR: borrow of moved value
}