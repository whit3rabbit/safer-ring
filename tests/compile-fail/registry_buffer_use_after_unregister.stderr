error[E0382]: borrow of moved value: `registered_buffer`
  --> tests/compile-fail/registry_buffer_use_after_unregister.rs:13:33
   |
7  |     let registered_buffer = registry.register_buffer(buffer).unwrap();
   |         ----------------- move occurs because `registered_buffer` has type `RegisteredBuffer`, which does not implement the `Copy` trait
...
10 |     let _returned_buffer = registry.unregister_buffer(registered_buffer).unwrap();
   |                                                       ----------------- value moved here
...
13 |     println!("Buffer size: {}", registered_buffer.size()); //~ ERROR: borrow of moved value
   |                                 ^^^^^^^^^^^^^^^^^ value borrowed here after move
